### ISSUE TITLE
[Question] How to have ConanFile.run() use a different Python version

### ISSUE THREAD

**Issue by Adnn** [conan-io/conan#4988]
We are following the recommendation of installing Conan on Python 3.

Yet, we are now facing a complication: one of our dependency (Qt 5.9 web engine module) require Python 2.7 to build on Windows.
Usually, when we need an external tool, we can rely on a (usually already existing) "xxx_installer" package, like `cmake_installer/3.12.0@conan/stable`. Sadly, it seems that changing the Python interpreter would be a bit more involved, as the command already executes within a Python context.

Is there a way to run the `ConanFile.run()` command with a different Python interpreter? Is it possible to handle that as a `build_requirement` ?

**Comment by memsharded** [comment]
Hi @Adnn 

It depends on what you want to achieve, and how Qt5.9 web engine module uses python. If internally it can call ``python2 ...``, then there is not a problem, as it runs as a subprocess, it can run another python version if instructed to. If the script needs to run in the same python that is running Conan, then it might be impossible. But I wouldn't say so, all package builds should be subprocess (``self.run()`` launch a subprocess), it doesn't matter if it is cmake, make, or python. 

A python27 installer package might help, but only if the path or the environment that Qt is using to launch that python is compatible.  Please tell us a bit more about that, thanks!

**Comment by Adnn** [comment]
Hi @memsharded, and thank you for your response,

I don't know how Qt web engine is calling python, this is somewhere in a gigantic chromium codebase with other tools like Ninja taking care of some steps... I would not venture my poor soul there ; )
What I do know is that if Conan runs on Python 2.7, the build succeeds, but if it runs Python 3.7, the build fails with this error (emphasis mine):

> bootstrapping ninja...
> warning: A compatible version of re2c (>= 0.11.3) was not found; changes to src/*.in.cc will not affect your build.
> wrote build.ninja.
> bootstrap complete. rebuilding...
> C:\Users\adn.conan\data\jom_installer\1.1.2\bincrafters\stable\package\3475bd55b91ae904ac96fde0f106a136ab951a5e\bin\jom.exe -f Makefile.ninja.Release
> ( if not exist Makefile.gn C:.conan\9b7700\1\qtbase\bin\qmake.exe -o Makefile.gn C:.conan\543f17\1\qt5\qtwebengine\src\buildtools\gn.pro ) && C:\Users\adn.conan\data\jom_installer\1.1.2\bincrafters\stable\package\3475bd55b91ae904ac96fde0f106a136ab951a5e\bin\jom.exe -f Makefile.gn
> File "C:.conan\543f17\1\qt5\qtwebengine\src\3rdparty\chromium\tools\gn\bootstrap\bootstrap.py", line 72
**> print 'Building gn manually in a temporary directory for bootstrapping...'
> ^
> SyntaxError: Missing parentheses in call to 'print'. Did you mean print('Building gn manually in a temporary directory for bootstrapping...')?**
> Project ERROR: GN build error!
> jom: C:.conan\9b7700\1\qtwebengine\src\buildtools\Makefile [sub-gn-pro-make_first] Error 3
> jom: C:.conan\9b7700\1\qtwebengine\src\Makefile [sub-buildtools-make_first] Error 2
> jom: C:.conan\9b7700\1\qtwebengine\Makefile [sub-src-make_first] Error 2
> jom: C:.conan\9b7700\1\Makefile [module-qtwebengine-make_first] Error 2
> Qt/5.9.7@bincrafters/stable:
> Qt/5.9.7@bincrafters/stable: ERROR: Package 'c14495df58cab94290b555ce47f861cbfaf761bf' build failed
> Qt/5.9.7@bincrafters/stable: WARN: Build folder C:.conan\9b7700\1
> ERROR: Qt/5.9.7@bincrafters/stable: Error in build() method, line 348
> _build("jom")
> while calling '_build', line 342
> self.run(make)
> ConanException: Error 2 while executing jom

So, it looks like Python is running, so the build process is probably simply calling python with un-versioned `python ...`.

What my recipe does is to call the actual build command (`configure` then `jom` on Windows) in a `self.run()`. Then, it is very likely that the jom process somewhere calls scripts who in turn call `python`. Is there a way to have this `self.run()` execute in an environment where Python 2.7 would be found before the current Python 3.7?

**Comment by Adnn** [comment]
PS: Is someone aware if a Python27 installer package is publicly available somewhere? There does not seems to be one in conan center.

**Comment by memsharded** [comment]
> What my recipe does is to call the actual build command (configure then jom on Windows) in a self.run(). Then, it is very likely that the jom process somewhere calls scripts who in turn call python. Is there a way to have this self.run() execute in an environment where Python 2.7 would be found before the current Python 3.7?

> PS: Is someone aware if a Python27 installer package is publicly available somewhere? There does not seems to be one in conan center.

I think that is the issue, even if you have a Python27 conan installer package, it won't be automatically used, because it seems that somewhere inside the toolchain, it is using the python that it finds in the path, that it is the current one.

I have tried this (to execute Py37 from a Py27 running Conan one) and it worked well:

```python

    def build(self):
        with tools.environment_append({"PATH": "C:\Python37-32"}):
            self.run("python --version")
```
Prints 
```
Python 3.7.0
```

So the opposite is totally doable too. I'd recommend trying that, with a Py27 installed in your system first, then try it with a Conan installer package. Please tell me how it goes, thanks!

**Comment by Adnn** [comment]
Thank you @memsharded, the `tools.environment_append` solution indeed allowed the build-chain to use Python 2.7, and the build succeeded.

Your remark regarding installer package raises a new question for me though:
> [...] even if you have a Python27 conan installer package, it won't be automatically used, because it seems that somewhere inside the toolchain, it is using the python that it finds in the path, that it is the current one.

I thought it was one of the feature of installer packages to obtain a tool binary, then to make sure this binary is accessible in the path?

**Comment by memsharded** [comment]
> I thought it was one of the feature of installer packages to obtain a tool binary, then to make sure this binary is accessible in the path?

Yes, yes, that is totally possible, you can have an installer package for python2, that injects itself in the PATH environment variable. 

What I meant is that if the Qt had some internal logic of locating the python executable, lets say that it could check if you are already in a python virtualenv, and decide to run it instead of the one that is found first in the PATH, then, that would be difficult to fix without patching the Qt scripts.

**Comment by Adnn** [comment]
Just to confirm, making a naïve Python2.7 installer package allows to build Qt when Conan runs on Python 3. Thank you for the help in achieving that!

As a last question, is someone aware of any Python 2.7 installer package maintained in one of the Conan public repositories? I'd be very happy to move away from my ... hackish ... recipe ; )
